1710: Implement SELinuxChangePolicy

jsafrane opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

* Add PodSecurityContext.`SELinuxChangePolicy` field, with its own `SELinuxChangePolicy` feature gate.
* Implement the field usage in kubelet

Short summary of the KEP:
Enabling `SELinuxMount` feature gate can break some Pods. Therefore we offer a proactive opt-out using `SELinuxChangePolicy` field + feature gate.
* When `SELinuxChangePolicy` is enabled and `SELinuxMount` still disabled, the new field only accepts value `Recursive`, but does not act on it except for RWOP volumes.
    * The only visible result on non-RWOP volumes is that `volume_manager_selinux_volume_context_mismatch_warnings_total` is not bumped when two pods with different SELinux labels and both with `Recursive` policy use the same volume in parallel.
    * This allows users to enable `SELinuxChangePolicy`, see what Pods would be broken *before* `SELinuxMount` is enabled, update `SELinuxChangePolicy` of these Pods and enable `SELinuxMount` only when all Pods are safe and won't break.
* When `SELinuxMount` gets enabled, `SELinuxChangePolicy` field gets effective and `Recursive` means the Pod opts out from mounting with -o context, even when the CSI driver supports it.
    * This allows two pods with different SELinux labels and both with `Recursive` policy use the same volume in parallel. Typically a privileged one (with `spc_t`) and unprivileged one (with `cX,cY`, where X and Y are random numbers).
* Validation of the field is ... non-standard.
  * If `SELinuxChangePolicy` is enabled and `SELinuxMount` disabled, the only allowed value is `Recursive` (or nil).
  * If `SELinuxChangePolicy` is enabled and `SELinuxMount` enabled, both `MountOption` and `Recursive` (and nil) are allowed.
  * Usual ratcheting validation should apply - an old object with any value is valid, regardless what values are allowed with current feature gate combination.
  * For the record, PodSecurityContext cannot be edited after pod creation, so there is no field update validation.

I had to rework some unit tests to use table approach, the number of unit tests grows a lot with different access modes,  SELinuxChange policies, SELinux options and feature gates enabled. Still, the actual changes in kubelet should be relatively small.

TODO in subsequent PRs:
* Add e2e tests for the new field, both with `SELinuxMount` feature gate disabled and enabled
* Add a CI job with `SELinuxChangePolicy` enabled and `SELinuxMount` disabled. 
* Implement the new controller proposed in the KEP. The controller will just emit metrics + events, it won't affect volume or pod lifecycle. All parts of the feature that affects kubelet and the API server are already here.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Implemented a new, alpha `seLinuxChangePolicy` field within a Pod-level `securityContext`, under SELinuxChangePolicy feature gate. This field allows for opting out from mounting Pod volumes with SELinux label when SELinuxMount feature is enabled (it is alpha and disabled by default now).
Please see [the KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling#story-3-cluster-upgrade) how we expect to warn users before any SELinux behavior changes and how they can opt-out before. Note that this field and feature gate is useful only with clusters that run with SELinux enabled. No action is required on clusters without SELinux.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1710-selinux-relabeling
```

jsafrane committed:
Introduce SELinuxChangePolicy

Feature gate + the API

jsafrane committed:
Regenerated API

jsafrane committed:
Add SELinuxChangePolicy validation

jsafrane committed:
Use SELinuxChangePolicy

jsafrane committed:
Refactor getSELinuxLabel

Return early and reduce indentation

jsafrane committed:
Refactor AddPodToVolume unit tests with a single pod

Refactor existing tests that run AddPodToVolume with a single pod / volume
to a table. And add few extra tests, covering existing ReadWriteOncePod
functionality.

jsafrane committed:
Refactor AddPodToVolume unit tests with multiple pods

Refactor existing tests that run AddPodToVolume with multiple pods
to a table.

jsafrane committed:
Add SELinuxChangePolicy tests

jsafrane committed:
Add SELinuxMount tests

jsafrane commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -407,14 +407,28 @@ func (dsw *desiredStateOfWorld) getSELinuxLabel(volumeSpec *volume.Spec, seLinux
 				newLabel, err := dsw.seLinuxTranslator.SELinuxOptionsToFileLabel(containerContext)
 				if err != nil {
-					fullErr := fmt.Errorf("failed to construct SELinux label from context %q: %s", containerContext, err)
+					fullErr := fmt.Errorf("failed to construct SELinux label from context %q: %w", containerContext, err)
```
this was requested by a go linter

jsafrane commented:
/retest

jsafrane commented:
/assign @msau42 @gnufied 
for review

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -3813,6 +3829,15 @@ type PodSecurityContext struct {
 	// Note that this field cannot be set when spec.os.name is windows.
 	// +optional
 	AppArmorProfile *AppArmorProfile
+	// seLinuxChangePolicy defines how the container's SELinux label is applied to all volumes used by the Pod.
```
Can multiple pods specify different policies?

jsafrane replied:
They can, as long as they don't share volumes.

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -3677,6 +3677,22 @@ const (
+type PodSELinuxChangePolicy string
+
+const (
+	// Recursive relabeling of all Pod volumes by the container runtime.
```
I think these comments do not end up in the generated api docs, so it would be better to put these in the comment for the SELinuxChangePolicy field.

jsafrane replied:
Added. It's getting bigger and bigger :-)

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -3813,6 +3829,15 @@ type PodSecurityContext struct {
+	// It affects only in-tree iSCSI and FibreChannel volumes, and CSI volumes that announce "SELinuxMount: true" in their CSIDriver instance.
+	// It affects only Pods that have SELinux label set, either in PodSecurityContext or in SecurityContext of all containers.
+	// Note that this field cannot be set when spec.os.name is windows.
+	// +featureGate=SELinuxChangePolicy
```
Can you also document that `MountOption` is only allowed if the SELinuxMount feature gate is enabled?

jsafrane replied:
added

msau42 reviewed (commented):
> This allows users to enable SELinuxChangePolicy, see what Pods would be broken before SELinuxMount is enabled

Can you clarify how this users can tell it is broken?

BenTheElder commented:
/remove-sig api-machinery
/sig architecture

jsafrane commented:
/label api-review

gnufied commented on pkg/apis/core/types.go:
```diff
@@ -3813,6 +3829,28 @@ type PodSecurityContext struct {
 	AppArmorProfile *AppArmorProfile
+	// seLinuxChangePolicy defines how the container's SELinux label is applied to all volumes used by the Pod.
+	// It has no effect on nodes that do not support SELinux or to volumes does not support SELinux.
+	// Valid values are "MountOption" and "Recursive". If not specified, "MountOption" is used.
```
>If not specified, "MountOption" is used.
Not yet right? I am not sure if the nuanced way this is being implemented needs to be documented in API docs. But this certainly feels incorrect here technically.

gnufied commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -392,62 +392,66 @@ func (dsw *desiredStateOfWorld) AddPodToVolume(
-func (dsw *desiredStateOfWorld) getSELinuxLabel(volumeSpec *volume.Spec, seLinuxContainerContexts []*v1.SELinuxOptions) (string, bool, error) {
-	var seLinuxFileLabel string
-	var pluginSupportsSELinuxContextMount bool
+func (dsw *desiredStateOfWorld) getSELinuxLabel(volumeSpec *volume.Spec, seLinuxContainerContexts []*v1.SELinuxOptions, podSecurityContext *v1.PodSecurityContext) (string, bool, error) {
```
Can we document what this function returns? It was bit hard to read.

jsafrane replied:
Added a comment.

gnufied commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -392,62 +392,66 @@ func (dsw *desiredStateOfWorld) AddPodToVolume(
+		return "", pluginSupportsSELinuxContextMount, nil
+	}
+
+	seLinuxSupported := util.VolumeSupportsSELinuxMount(volumeSpec)
```
Why does `VolumeSupportsSELinuxMount` return false for volumes without access modes?

jsafrane replied:
Good catch! That's leftover of RWOP, which needed to be the only item in AccessModes.

gnufied commented on pkg/kubelet/volumemanager/cache/desired_state_of_world.go:
```diff
@@ -301,7 +301,7 @@ func (dsw *desiredStateOfWorld) AddPodToVolume(
 	}
 
-	seLinuxFileLabel, pluginSupportsSELinuxContextMount, err := dsw.getSELinuxLabel(volumeSpec, seLinuxContainerContexts)
+	seLinuxFileLabel, pluginSupportsSELinuxContextMount, err := dsw.getSELinuxLabel(volumeSpec, seLinuxContainerContexts, pod.Spec.SecurityContext)
```
Not visible in this PR, but should `GetPodVolumeNames` be changed to return selinux context for newer featuregate as well? Currently I think it will only return `seLinuxContainerContexts` for `ReadWriteOnce` feature-gate enabled.

jsafrane replied:
`GetPodVolumeNames` returns all pod volumes and all their labels, regardless of `SELinuxMount` feature gate.

Caller of `GetPodVolumeNames` must check the gates to produce the right errors / warnings / metrics if the access mode + feature gate do not match.

jsafrane replied:
Ok, I moved it out. Most `GetPodVolumeNames` don't need to collect SELinuxOptions and those who do, they can evaluate the feature gate before the call.

jsafrane committed:
Document getSELinuxLabel function

jsafrane committed:
Fix access mode evaluation

Now that SELinuxMount can support volumes with any access modes, length of
the access modes does not need to be 1. That was requirement for RWOP
volumes only.

fedebongio commented:
/remove-sig api-machinery

jsafrane committed:
Collect SELinux options only when needed

Remove feature gate check from GetPodVolumeNames and collect SELinux
options only when it's really needed.

jsafrane committed:
Remove useless comment

seLinuxOptions are copied few lines below during `append()`

jsafrane commented:
/retest
TestHealthEndpoints//readyz_with_sched-handler-sync flaked

cici37 commented:
/remove-sig api-machinery

msau42 commented on pkg/apis/core/types.go:
```diff
@@ -3813,6 +3829,32 @@ type PodSecurityContext struct {
+	// CSIDriver instance. Other volumes are always re-labelled recursively.
+	// "MountOption" value is allowed only when SELinuxMount feature gate is enabled.
+	//
+	// If not specified and SELinuxMount feature gate is enabled, "MountOption" is used.
```
I understand that it will make catching impacted workloads easier, but changing the default behavior based on which feature gates are enabled seems like a confusing user experience. Would it be safer to just have workloads opt-in on their own?

jsafrane replied:
We discussed that with @deads2k. Our numbers show 1.15% of OpenShift clusters have at least one pod (Deployment, StatefulSet) that may require opt-out. 98.85%  of cluster won't need to do anything and can just upgrade.

(Rounding up), we considered that it makes more sense to let the 2% to force them to opt-out (and only few pods in these clusters need to opt-out) than 98% to opt-in all their pods.

jsafrane replied:
We discussed this with @deads2k. Current OpenShift numbers show that 1.15% of clusters have at least one Pod (Deployment, StatefulSet) that would need an opt-out. 98.85% of clusters can just upgrade and get faster relabeling. 

Exaggerating the number, we think it's better to warn 2% of clusters to opt-out before upgrade, than let 98% of clusters to opt-in explicitly.

msau42 replied:
Ok, can you update the release notes with some more warnings and/or a link to a page where there's more information on how they can test/check?

jsafrane replied:
Edited

jsafrane commented:
rebased

gnufied commented:
/lgtm
/approve

msau42 reviewed (commented):
/lgtm
/approve

msau42 commented:
/assign @jpbetz

jpbetz commented:
/approve
For apply configurations

tengqm commented:
Do we have a PR for this feature (gate) on website?

fedebongio commented:
/triage accepted
